README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

HDLFPCA
=======

<!-- badges: start -->

[![R build
status](https://github.com/seonjoo/HDLFPCA//workflows/R-CMD-check/badge.svg)](https://github.com/seonjoo/HDLFPCA//actions)
<!-- badges: end -->

The goal of HDLFPCA is to Perform High-dimensional Longitudinal PCA.

Installation
------------

You can install the development version from
[GitHub](https://github.com/) with:

    # install.packages("devtools")
    devtools::install_github("seonjoo/HDLFPCA")

Usage
-----

    library(HDLFPCA)
    example_hd_data = generate_hd_data(I = 100)
    phix0 = example_hd_data$phix0
    phix1 = example_hd_data$phix1
    phiw = example_hd_data$phiw
    Y = example_hd_data$Y
    time = example_hd_data$time
    J = example_hd_data$J
    I = example_hd_data$I
    visit = example_hd_data$visit
    re <- hd_lfpca(
        Y,
        T = scale(time, center = TRUE, scale = TRUE),
        J = J,
        I = I,
        visit = visit,
        varthresh = 0.95,
        projectthresh = 1,
        timeadjust = FALSE,
        figure = TRUE
      )
    #> Reduce Dimension
    #> Estimate Covariance Functions
    #> Calcuate Scores
    #> Compute Residual to the Demeaned Data.
    #> Residual of LFPCA Model is: 9872.67284871758

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

    cor(phix0, re$phix0)
    #>             [,1]        [,2]        [,3]        [,4]
    #> [1,]  0.99720326  0.09083731  0.02831191 -0.03210385
    #> [2,]  0.04360455 -0.99192254  0.03077443  0.07037896
    #> [3,] -0.02504620  0.01250388 -0.97917094  0.11347137
    cor(phix1, re$phix1)
    #>             [,1]        [,2]         [,3]       [,4]
    #> [1,]  0.99755948  0.08022781 -0.001390277 0.12272830
    #> [2,]  0.04494263 -0.98371159  0.029040865 0.05042425
    #> [3,] -0.02278886  0.01911020 -0.993862522 0.07308564

    library(gplots)
    #> 
    #> Attaching package: 'gplots'
    #> The following object is masked from 'package:stats':
    #> 
    #>     lowess
    par(mfrow = c(2, 2),
        mar = rep(0.5, 4),
        bg = "gray")
    bs = c(-100:100) / 1000 * 1.5

    image(phix0,
          axes = FALSE,
          col = gplots::bluered(200),
          breaks = bs)
    image(re$phix0[, 1:3],
          axes = FALSE,
          col = gplots::bluered(200),
          breaks = bs)
    image(phix1,
          axes = FALSE,
          col = gplots::bluered(200),
          breaks = bs)
    image(re$phix1[, 1:3],
          axes = FALSE,
          col = gplots::bluered(200),
          breaks = bs)

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" />


NEWS.md
# HDLFPCA 1.2.0

* Added a `NEWS.md` file to track changes to the package.
* Added `example_hd_data` and `example_hd_data_small` data sets for package testing.
* Added tests


tests/testthat.R
library(testthat)
library(HDLFPCA)

test_check("HDLFPCA")


tests/testthat/test-full_dataset.R
testthat::context("Trying HDLFPCA")
example_hd_data = generate_hd_data(I = 100)

phix0 = example_hd_data$phix0
phix1 = example_hd_data$phix1
phiw = example_hd_data$phiw
Y = example_hd_data$Y
time = example_hd_data$time
J = example_hd_data$J
I = example_hd_data$I
visit = example_hd_data$visit

testthat::test_that("works on full data", {

  indices = 1:nrow(Y)
  re <- HDLFPCA::hd_lfpca(
    Y[indices,],
    T = scale(time, center = TRUE, scale = TRUE),
    J = J,
    I = I,
    visit = visit,
    varthresh = 0.95,
    projectthresh = 1,
    timeadjust = FALSE
  )
  testthat::expect_equal(mean(re$phiw*1000), 0.0221596715582231)
  testthat::expect_equal(
    colMeans(re$phix0*1000),
    c(2.25835618679043, -2.21949537547265, -2.42342116795384, -2.88944204581535
    ))
  testthat::expect_equal(colMeans(re$phix1*1000),
                         c(1.90243315846835, -2.08446163591179,
                           -1.62285266405034, -0.634547910061792)
  )
  testthat::expect_equal(re$Nw, 7L)
  testthat::expect_equal(re$Nx, 4L)

  testthat::expect_equal(
    re$sx,
    c(490.77704840249, 96.8782474283866, 38.5230100680366, 4.5204046404315
    )
  )
  testthat::expect_equal(
    re$sw,
    c(114.551035546181, 30.3927525541092, 12.6548160599522, 11.9475847172898,
      8.36265901505133, 5.29074435789924, 3.63804451590994)
  )

  testthat::expect_equal(re$residual, 9872.67284871758)
  testthat::expect_equal(
    rowMeans(re$xi),
    c(0.677984623456904, -0.410937698650476, 0.16902003091013, 0.0194506415030195)
  )
})



testthat::test_that("works on subset data", {
  indices = 1:400
  re <- HDLFPCA::hd_lfpca(
    Y[indices,],
    T = scale(time, center = TRUE, scale = TRUE),
    J = J,
    I = I,
    visit = visit,
    Nw = 7,
    Nx = 4,
    varthresh = 0.95,
    projectthresh = 1,
    timeadjust = TRUE,
    figure = TRUE
  )


  testthat::expect_equal(mean(re$phiw*1000), -2.63946792438721)
  testthat::expect_equal(
    colMeans(re$phix0*1000),
    c(13.3493723861197, -13.3580732255287,
      -14.2666551815559, -18.5168153134067
    )
  )
  testthat::expect_equal(
    colMeans(re$phix1*1000),
    c(12.9816278713505, -13.9120079554871,
      -10.9841124093542, -10.2772410142864
    )
  )
  testthat::expect_equal(re$Nw, 7L)
  testthat::expect_equal(re$Nx, 4L)

  testthat::expect_equal(
    re$sx,
    c(559.122970613659, 110.110616086869, 43.7670421892069, 4.30591146765645
    )
  )
  testthat::expect_equal(
    re$sw,
    c(114.444177475096, 30.295055513728, 12.4159339484643, 11.7040242553597,
      5.021148985976, 3.14593397314224, 0.0513151608548925)
  )

  testthat::expect_equal(re$residual, 4132.61295466348)
  testthat::expect_equal(
    rowMeans(re$xi),
    c(0.75594903527378, -0.218963018530693, 0.194753903278344, 0.0231163909575934
    )
  )

})


testthat::test_that("Added in verbose and Nx > 20", {
  indices = 1:400
  re <- HDLFPCA::hd_lfpca(
    Y[indices,],
    T = scale(time, center = TRUE, scale = TRUE),
    J = J,
    I = I,
    visit = visit,
    Nw = 7,
    Nx = 20,
    varthresh = 0.95,
    projectthresh = 1,
    timeadjust = TRUE,
    figure = TRUE,
    verbose = 2
  )



})



R/generate_hd_data.R
#' Generate High-Dimensional Data
#'
#' @param I number of subjects
#'
#' @export
#'
#' @return A list with 7 elements, which are:
#' \describe{
#' \item{Y}{p-by-J matrix of values}
#' \item{I}{Total number of subjects}
#' \item{J}{Total number of observations}
#' \item{time}{Time of the image collection}
#' \item{visit}{Vector of number of visits per subjects}
#' \item{phix0}{a matrix of true phi values for x0}
#' \item{phix1}{a matrix of truephi values for x1}
#' \item{phiw}{a matrix of true phi values for w}
#' \item{xi}{covariate values}
#' \item{zeta}{true zeta}
#' }
#'
#' @examples
#' generate_hd_data()
#' generate_hd_data(I = 20)
generate_hd_data = function(I = 100) {
  set.seed(12345678)
  visit = stats::rpois(I, 1) + 3
  time = unlist(lapply(visit, function(x)
    scale(c(0, cumsum(
      stats::rpois(x - 1, 1) + 1
    )))))
  J = sum(visit)
  V = 2500
  phix0 = matrix(0, V, 3)
  phix0[1:50, 1] <- .1

  phix0[1:50 + 50, 2] <- .1

  phix0[1:50 + 100, 3] <- .1
  phix1 = matrix(0, V, 3)

  phix1[1:50 + 150, 1] <- .1

  phix1[1:50 + 200, 2] <- .1

  phix1[1:50 + 250, 3] <- .1
  phiw = matrix(0, V, 3)

  phiw[1:50 + 300, 1] <- .1

  phiw[1:50 + 350, 2] <- .1

  phiw[1:50 + 400, 3] <- .1
  xi = t(matrix(stats::rnorm(I * 3), ncol = I) * c(8, 4, 2)) * 3
  zeta = t(matrix(stats::rnorm(J * 3), ncol = J) * c(8, 4, 2)) * 2
  Y = phix0 %*% t(xi[rep(1:I, visit), ]) +
    phix1 %*% t(time * xi[rep(1:I, visit), ]) +
    phiw %*% t(zeta) + matrix(stats::rnorm(V * J, 0, .1), V, J)

  list(
    I = I,
    J = J,
    V = V,
    phix0 = phix0,
    phix1 = phix1,
    phiw = phiw,
    xi = xi,
    zeta = zeta,
    Y = Y,
    time = time,
    visit = visit
  )
}


data-raw/example_hd_data.R
## code to prepare `DATASET` dataset goes here
set.seed(12345678)
I = 100
visit = rpois(I, 1) + 3
time = unlist(lapply(visit, function(x)
  scale(c(0, cumsum(
    rpois(x - 1, 1) + 1
  )))))
J = sum(visit)
V = 2500
phix0 = matrix(0, V, 3)
phix0[1:50, 1] <- .1

phix0[1:50 + 50, 2] <- .1

phix0[1:50 + 100, 3] <- .1
phix1 = matrix(0, V, 3)

phix1[1:50 + 150, 1] <- .1

phix1[1:50 + 200, 2] <- .1

phix1[1:50 + 250, 3] <- .1
phiw = matrix(0, V, 3)

phiw[1:50 + 300, 1] <- .1

phiw[1:50 + 350, 2] <- .1

phiw[1:50 + 400, 3] <- .1
xi = t(matrix(rnorm(I * 3), ncol = I) * c(8, 4, 2)) * 3
zeta = t(matrix(rnorm(J * 3), ncol = J) * c(8, 4, 2)) * 2
Y = phix0 %*% t(xi[rep(1:I, visit), ]) +
  phix1 %*% t(time * xi[rep(1:I, visit), ]) +
  phiw %*% t(zeta) + matrix(rnorm(V * J, 0, .1), V, J)

example_hd_data = list(
  I = I,
  J = J,
  V = V,
  phix0 = phix0,
  phix1 = phix1,
  phiw = phiw,
  xi = xi,
  zeta = zeta,
  Y = Y,
  time = time,
  visit = visit
)
usethis::use_data(example_hd_data, overwrite = TRUE, compress = "xz")


data-raw/example_hd_data_small.R
## code to prepare `DATASET` dataset goes here
set.seed(12345678)
I = 20
visit = rpois(I, 1) + 3
time = unlist(lapply(visit, function(x)
  scale(c(0, cumsum(
    rpois(x - 1, 1) + 1
  )))))
J = sum(visit)
V = 2500
phix0 = matrix(0, V, 3)
phix0[1:50, 1] <- .1

phix0[1:50 + 50, 2] <- .1

phix0[1:50 + 100, 3] <- .1
phix1 = matrix(0, V, 3)

phix1[1:50 + 150, 1] <- .1

phix1[1:50 + 200, 2] <- .1

phix1[1:50 + 250, 3] <- .1
phiw = matrix(0, V, 3)

phiw[1:50 + 300, 1] <- .1

phiw[1:50 + 350, 2] <- .1

phiw[1:50 + 400, 3] <- .1
xi = t(matrix(rnorm(I * 3), ncol = I) * c(8, 4, 2)) * 3
zeta = t(matrix(rnorm(J * 3), ncol = J) * c(8, 4, 2)) * 2
Y = phix0 %*% t(xi[rep(1:I, visit), ]) +
  phix1 %*% t(time * xi[rep(1:I, visit), ]) +
  phiw %*% t(zeta) + matrix(rnorm(V * J, 0, .1), V, J)

example_hd_data_small = list(
  I = I,
  J = J,
  V = V,
  phix0 = phix0,
  phix1 = phix1,
  phiw = phiw,
  xi = xi,
  zeta = zeta,
  Y = Y,
  time = time,
  visit = visit
)
usethis::use_data(example_hd_data_small, overwrite = TRUE, compress = "xz")


R/hd_lfpca.R
#' Conduct high-dimensional longitudinal principal component analysis
#'
#'
#' For V-by-J dimensional matrix of Y, HDLPCA first reduce the dimension of p to J if V>J:
#' `Y = UDV`. Then, longitudinal PCA is applied to the projected scores of Y on the p-dimensional subspace: Y1=DY.
#' The dimension of the subspace is selected based on the variance explained by first p-components.
#' The default value is `projectthresh`=1.
#'
#' @param Y p-by-J matrix,
#' @param T Time of the image collection
#' @param J Total number of observations
#' @param I Total number of subjects
#' @param visit Vector of number of visits per subjects
#' @param verbose print diagnostic messages, higher number for more verbosity
#' @param Nx Dimension of the subject-specific components
#' @param Nw Dimension of the subject-visit specific components
#' @param varthresh (default=0.99) Threshold for variance explained for both subject-specific and subject-visit specific components for dimension selection
#' @param projectthresh Threshold for variance explain in the first step of SVD
#' @param timeadjust (default=FALSE) Scale time per subject
#' @param figure should a bar plot of the variance explained be plotted?
#' @return A list with the following elements:
#' \describe{
#' \item{xi}{Subject-specific principal component scores.}
#' \item{phix0}{Subject-specific principal component loading corresponding to intercept.}
#' \item{phix1}{Subject-specific principal component loading corresponding to slope.}
#' \item{zeta}{Subject-visit-specific principal component scores.}
#' \item{phiw}{Subject-visit-specific principal component loading}
#' }
#' @examples
#' run_list = generate_hd_data(I = 20)
#' phix0 = run_list$phix0
#' phix1 = run_list$phix1
#' phiw = run_list$phiw
#' Y = run_list$Y
#' time = run_list$time
#' J = run_list$J
#' I = run_list$I
#' visit = run_list$visit
#' # indices = 1:2500 # all indices
#' indices = 1:700
#' re <- HDLFPCA::hd_lfpca(
#'   Y[indices,],
#'   T = scale(time, center = TRUE, scale = TRUE),
#'   J = J,
#'   I = I,
#'   visit = visit,
#'   varthresh = 0.95,
#'   projectthresh = 1,
#'   timeadjust = FALSE
#'   )
#'
#' \donttest{
#' run_list = generate_hd_data(I = 100)
#' phix0 = run_list$phix0
#' phix1 = run_list$phix1
#' phiw = run_list$phiw
#' Y = run_list$Y
#' time = run_list$time
#' J = run_list$J
#' I = run_list$I
#' visit = run_list$visit
#' # indices = 1:2500 # all indices
#' indices = 1:700
#' re <- HDLFPCA::hd_lfpca(
#'   Y[indices,],
#'   T = scale(time, center = TRUE, scale = TRUE),
#'   J = J,
#'   I = I,
#'   visit = visit,
#'   varthresh = 0.95,
#'   projectthresh = 1,
#'   timeadjust = FALSE
#'   )
#' cor(phix0[indices,], re$phix0)
#' cor(phix1[indices,], re$phix1)
#' if (requireNamespace("gplots", quietly = TRUE)) {
#'   library(gplots)
#'   par(mfrow = c(2, 2),
#'       mar = rep(0.5, 4),
#'       bg = "gray")
#'   bs = c(-100:100) / 1000 * 1.5
#'   image(phix0[indices,],
#'         axes = FALSE,
#'         col = bluered(200),
#'         breaks = bs)
#'   image(re$phix0[, 1:3],
#'         axes = FALSE,
#'         col = bluered(200),
#'         breaks = bs)
#'   image(phix1[indices,],
#'         axes = FALSE,
#'         col = bluered(200),
#'         breaks = bs)
#'   image(re$phix1[, 1:3],
#'         axes = FALSE,
#'         col = bluered(200),
#'         breaks = bs)
#' }
#' }
#'
#' \donttest{
#' # need to transpose for SVD
#' re <- HDLFPCA::hd_lfpca(
#'   Y[1:400, ],
#'   T = scale(time, center = TRUE, scale = TRUE),
#'   J = J,
#'   I = I,
#'   visit = visit,
#'   Nw = 20,
#'   Nx = 16,
#'   varthresh = 0.95,
#'   projectthresh = 1,
#'   timeadjust = TRUE,
#'   verbose = 2,
#'   figure = TRUE
#' )
#' }
#' @author Seonjoo Lee, \email{sl3670@cumc.columbia.edu}
#' @references Zipunnikov, V., Greven, S., Shou, H., Caffo, B., Reich, D. S., & Crainiceanu, C. (2014). Longitudinal High-Dimensional Principal Components Analysis with Application to Diffusion Tensor Imaging of Multiple Sclerosis. The Annals of Applied Statistics, 8(4), 2175–2202.
#' @keywords hdlfpca glmnet
#' @import MASS
#' @importFrom stats var
#' @importFrom grDevices rainbow
#' @importFrom graphics barplot text title
#' @export
hd_lfpca = function(Y,
                    T,
                    J,
                    I,
                    visit,
                    verbose = 1,
                    # prefix=date(),
                    Nx = NA,
                    Nw = NA,
                    varthresh = 0.99,
                    projectthresh = 1,
                    timeadjust = FALSE,
                    figure = FALSE) {
  # library(MASS)
  #J: Total number visits out of all subjects (sum of visit)
  #T must be centered+scaled time information
  # ylist : list of subjects.
  ###
  ### Y = [Y1 Y2... YI] = VDU' (p by J matrix)
  ### Calculate U using iterative calculation of sum(Yi'Yi)
  #	Y_total_mean=apply(Y,1,mean)
  if (I != length(visit)) {
    warning("Visit should be length of subjects (I)")
  }
  n_visits = sum(visit)
  if (n_visits != J) {
    warning("Total number of visits should be equal to J")
  }
  if (n_visits != ncol(Y)) {
    warning("Number of visits should total number of observations (columns of Y)")
  }

  Y = t(scale(t(Y), center = TRUE, scale = FALSE)) # Y is centered.

  svdy = svd(t(Y) %*% Y) # There is a sign ambiguity between SVD and eigen function.
  if (dim(Y)[1] >= dim(Y)[2]) {
    if (verbose > 0) {
      message("Reduce Dimension")
    }
    N = J_projected = sum(cumsum(svdy$d ^ 2) / sum(svdy$d ^ 2) < projectthresh) #Dimensionality #
  } else{
    if (verbose > 0) {
      message("Do not Reduce Dimension")
    }
    #sum(cumsum(svdy$d^2)/sum(svdy$d^2)<projectthresh) #Dimensionality #
    N = J_projected = nrow(Y) - 1
  }

  if (verbose > 1) {
    message(paste0("Projected dimension: ", J_projected))
  }
  U = svdy$v[, 1:J_projected]
  D = diag(sqrt(svdy$d[1:J_projected]))
  S = diag(svdy$d[1:J_projected])

  V = Y %*% U %*% solve(D)
  Ynew = t(V[, 1:J_projected]) %*% Y #D %*% t(U)
  rm(list = c("svdy"))


  Yvec = matrix(0, J_projected ^ 2, sum(visit ^ 2))
  X = matrix(0, sum(visit ^ 2), 5)
  k = 0
  J_sq = 0
  for (j in 1:I) {
    #	  print(j)
    s = 1

    Ji = visit[j]

    Ti = T[(1:Ji) + k]

    #Ti=(Ti-mean(Ti))
    if (timeadjust == TRUE) {
      Ti = (Ti - mean(Ti)) / sqrt(var(Ti))
    }
    Yvec_i = matrix(0, J_projected * J_projected, Ji ^ 2)
    X_i = matrix(0, Ji * Ji, 5)
    Ytmp = Ynew[, (1:Ji) + k]
    for (j1 in 1:Ji) {
      for (j2 in 1:Ji) {
        Yvec_i[, s] = as.vector(Ytmp[, j1] %*% t(Ytmp[, j2]))
        X_i[s, ] = c(1, Ti[j2], Ti[j1], Ti[j1] * Ti[j2], as.numeric(j1 ==
                                                                      j2))
        s = s + 1
      }
    }
    Yvec[, (1:(visit[j] ^ 2)) + J_sq] = Yvec_i

    X[(1:visit[j] ^ 2) + J_sq, ] = X_i
    J_sq = J_sq + Ji ^ 2

    k = k + Ji

  }
  if (verbose > 0) {
    message("Estimate Covariance Functions")
  }

  #system.time(beta<- Yvec %*% X %*% chol2inv(chol(t(X)%*%X)))
  beta <- Yvec %*% X %*% solve(t(X) %*% X)
  #sum((beta-beta2)^2)
  K00 = matrix(beta[, 1], J_projected, J_projected)
  K01 = matrix(beta[, 2], J_projected, J_projected)
  K10 = matrix(beta[, 3], J_projected, J_projected)
  K11 = matrix(beta[, 4], J_projected, J_projected)
  Kw = matrix(beta[, 5], J_projected, J_projected)
  Kx = rbind(cbind(K00, K01), cbind(K10, K11))

  Ax = eigen(Kx)
  Aw = eigen(Kw)
  #Estimate the components
  if (is.na(Nx) | is.na(Nw)) {
    lim = 1
    Nw0 = Nx0 = Nw = Nx = min(J_projected,
                              sum(Ax$values > 0),
                              sum(Aw$values > 0))

    while (lim > varthresh) {
      if (min(Aw$values[1:Nw]) < min(Ax$values[1:Nx])) {
        Nw = Nw0 - 1
      } else{
        Nx = Nx0 - 1
      }
      lim = (sum(Ax$values[1:Nx]) + sum(Aw$values[1:Nw])) /
        (sum(Ax$values[1:sum(Ax$values > 0)]) +
           sum(Aw$values[1:sum(Aw$values > 0)]))

      Nx0 <- Nx
      Nw0 <- Nw
      #	if (lim>varthresh){Nx=Nx-1;Nw=Nw-1}
      if (verbose > 1) {
        message(paste(Nx, Nw, lim))
      }
    }
  }
  Ax0 =  Ax$vectors[1:J_projected, 1:Nx]
  Ax1 = Ax$vectors[1:J_projected + J_projected, 1:Nx]
  S_x = Ax$values[1:Nx]
  S_u = Aw$values[1:Nw]
  Au =  Aw$vectors[, 1:Nw]


  phix0 = V %*% Ax0
  phix1 = V %*% Ax1
  phiw = V %*% Aw$vectors[, 1:Nw]

  #rm(list=c("Ax","Aw"))

  #ramp = colorRamp(c("blue","gray","red"))
  #colscheme = rgb( ramp(seq(0, 1, length = 240)), max = 255)#redgreed(255)#
  b = grDevices::rainbow(200)
  phix = rbind(phix0, phix1)
  #(diag(t(phix) %*% phix))
  #(diag(t(phiw) %*% phiw))
  #(diag(t(phix0) %*% phix0))
  #(diag(t(phix1) %*% phix1))
  total_lambda = sum(Ax$values[1:Nx]) + sum(Aw$values[1:Nw])
  a = (diag(t(phix0) %*% phix0))

  if (figure == TRUE) {
    x = cbind(
      Baseline = Ax$values[1:Nx] / total_lambda * a * 100,
      Longitudinal = Ax$values[1:Nx] / total_lambda * (1 - a) * 100
    )
    maxval = max(apply(x, 1, sum))

    if (Nx < 15) {
      h = barplot(
        t(x[1:Nx, ]),
        border = b[c(100, 170)],
        beside = FALSE,
        col = b[c(100, 170)],
        names.arg = 1:Nx,
        legend = TRUE,
        ylim = c(0, 1.1 * maxval)
      )
      title("Variability Explained by subject-specific components")
      text(h[1:Nx],
           Ax$values[1:Nx] / total_lambda * 100 + 0.06 * maxval,
           paste(round(Ax$values[1:Nx] / total_lambda * 10000) / 100, '%', sep = ""))
      text(
        h[1:Nx],
        Ax$values[1:Nx] / total_lambda * 100 + 0.02 * maxval,
        paste("(", round((1 - a[1:Nx]) * 10000) / 100, '%)', sep = "")
      )
    }
    if (Nx >= 15) {
      h = barplot(
        t(x[1:15, ]),
        border = b[c(100, 170)],
        beside = FALSE,
        col = b[c(100, 170)],
        names.arg = 1:15,
        legend = TRUE,
        ylim = c(0, 1.1 * maxval)
      )
      title("Variability Explained by First 15 Eigenvectors - subject-specific variation")
      text(h[1:15],
           Ax$values[1:15] / total_lambda * 100 + 0.06 * maxval,
           paste(round(Ax$values[1:15] / total_lambda * 10000) / 100, '%', sep = ""))
      text(
        h[1:15],
        Ax$values[1:15] / total_lambda * 100 + 0.02 * maxval,
        paste("(", round((1 - a[1:15]) * 10000) / 100, '%)', sep = "")
      )
    }
  }

  #par(mfrow=c(2,5))
  #for (j in 1:5){image(matrix(phix0[,j],200,200),breaks=-120:120/2000,col=colscheme,)}
  #for (j in 1:5){image(matrix(phix1[,j],200,200),breaks=-120:120/2000,col=colscheme,)}

  ## Step 4. EBLUPs
  xi_est = matrix(0, Nx, I)


  zeta_est = matrix(0, Nw, J)


  C00 = t(Ax0) %*% Ax0

  C11 = t(Ax1) %*% Ax1

  C01 = t(Ax0) %*% Ax1

  C10 = t(C01)

  C0W = t(Ax0) %*% Au

  C1W = t(Ax1) %*% Au

  if (verbose > 0) {
    message("Calcuate Scores")
  }
  Ut = t(U)

  k = 0

  for (i in 1:I) {
    Ji = visit[i]

    Ti = T[k + 1:Ji]

    if (timeadjust == TRUE) {
      Ti = (Ti - mean(Ti)) / sqrt(var(Ti))
    }

    Ti_dot = sum(Ti)
    # it's 0
    Ti_sq_dot = sum(Ti ^ 2)
    # it's 1
    Ui = Ut[, k + 1:Ji]


    ones_Ji = rep(1, Ji)


    #   %%%% BiBi' %%%%
    BBi_11 = Ji * C00 + Ti_sq_dot * C11 + Ti_dot * (C10 + C01)

    BBi_12 = kronecker(t(ones_Ji), C0W) + kronecker(t(Ti), C1W)

    BBi_21 = t(BBi_12)

    BBi_22 = diag(rep(1, Nw * Ji))

    #       BBi = rbind(cbind(BBi_11, BBi_12),cbind(BBi_21, BBi_22))

    C1 = ginv(BBi_11 - BBi_12 %*% BBi_22 %*% BBi_21)
    C2 = BBi_22 + BBi_21 %*% ginv(BBi_11 - BBi_12 %*% BBi_21) %*% BBi_12
    C3 = ginv(BBi_11) %*% BBi_12 %*% C2

    #   %%%% Bi'Yi %%%%
    BYi_11 = as.vector(t(Ax0) %*% sqrt(S[1:N, ]) %*% Ui %*% ones_Ji + t(Ax1) %*%
                         sqrt(S[1:N, ]) %*% Ui %*% Ti)
    BYi_12 = as.vector(t(Au) %*% sqrt(S[1:N, ]) %*% Ui)


    BYi = c(BYi_11, BYi_12)

    #      system.time(wi<-chol2inv(chol(BBi))%*%BYi)
    wi <- rbind(cbind(C1, -C3), cbind(-t(C3), C2)) %*% BYi
    xi_est[, i] = wi[1:Nx]

    zeta_i_vec = wi[Nx + 1:(Ji * Nw)] # put the size
    for (j in 1:Ji) {
      zeta_est[, k + j] = zeta_i_vec[((j - 1) * Nw + 1):(j * Nw)]

    }
    k = k + Ji

  }


  if (verbose > 0) {
    message("Compute Residual to the Demeaned Data.")
  }
  tmpx = matrix(0, nrow(Ax0), J)
  cumvisit = 0
  for (j in 1:I) {
    for (k in 1:visit[j]) {
      tmpx[, cumvisit + k] = (Ax0 + Ax1 * T[cumvisit + k]) %*% xi_est[, j]
    }
    cumvisit = sum(visit[1:j])
  }
  residual = sum((Ynew - tmpx - Aw$vectors[, 1:Nw] %*% zeta_est) ^ 2)
  if (verbose > 0) {
    message(paste("Residual of LFPCA Model is:", residual))
  }
  result = new.env()
  result$xi = xi_est
  result$zeta = zeta_est
  result$phix0 = phix0
  result$phix1 = phix1
  result$Nx = Nx
  result$Nw = Nw
  result$phiw = phiw
  result$sx = Ax$values[1:Nx]
  result$sw = Aw$values[1:Nw]
  result$residual = residual
  result$tij = T
  result = as.list(result)
  return(result)
}
